### PR TITLE
std.cstr: deprecate namespace

### PR DESCRIPTION
**Pull request by BratishkaErik**

### CODE REVIEW

**Comment by andrewrk** [comment]
Can you provide some rationale for these changes? I personally prefer status quo.

**Comment by BratishkaErik** [comment]
> Can you provide some rationale for these changes? I personally prefer status quo.

addNullByte does exactly same thing as allocator.dupeZ (literally equal line-by-line), line_sep doesn't sounds as something that should belong in "c string" namespace, and NullTerminated2DArray is not used anywhere in repository.

**Comment by ifreund** [comment]
I like these changes, I'm not sure `cstr` has a good reason to exist tbh. Sentinel terminated pointers are a native zig type and functions to make working with them easier belong in `std.mem` in my opinion.

Perhaps it is reasonable to have a compatibility function matching libc's `strcmp`, but I think that would make more sense as `std.c.strcmp()` than the current `std.cstr.cmp()`.

Tangentially I think we may want an idiomatic Zig function for comparing sentinel terminated pointers as it could likely be faster than `std.mem.eql(u8, std.mem.sliceTo(a, 0), std.mem.sliceTo(b, 0))` in many cases where the "strings" are not equal and iteration to find the sentinel can be aborted.

**Comment by andrewrk** [comment]
OK, I actually agree with all of that. I think my only issue is with the new public namespace names chosen:

* `std.ascii.end_of_line.cr_lf`
* `std.ascii.end_of_line.lf`
* `std.ascii.end_of_line.native

Previously we had:

* `std.cstr.line_sep`

Looking at all the places `line_sep` was used, I think that API should actually be *eliminated*. In these places, the kind of newline used had more to do with the specifics of the callsite than on what target OS the zig code was being compiled for.

So, I am in favor of these changes, provided that this additional thing is done:

 * Remove `std.ascii.end_of_line` entirely.

A follow-up commit to fully eliminate std.cstr would be appreciated.

File: lib/std/c.zig [hunk 1]
```diff
@@ -39,6 +39,33 @@ pub fn versionCheck(comptime glibc_version: std.SemanticVersion) type {
     };
 }
 
+/// Compatibility function, used by Zig's implementation of libc in `lib/c.zig`.
+/// Assumes that `a` and `b` are multi-item pointers with NUL-termination.
+/// Returns `-1` if `a` appears before `b` in lexicographical order.
+/// Returns `1` if `a` appears after `b` in lexicographical order.
+/// Returns `0` if `a` and `b` are equal.
+/// See also `std.ascii.orderIgnoreCase` for idiomatic function
+/// with parameters as Zig's slices.
```
Comment (ifreund) [inline lib/std/c.zig:48]: This comment is incorrect, `std.mem.order()` is the idiomatic Zig equivalent.

`std.ascii.orderIgnoreCase()` has different behavior.

File: lib/std/c.zig [hunk 2]
```diff
@@ -39,6 +39,33 @@ pub fn versionCheck(comptime glibc_version: std.SemanticVersion) type {
     };
 }
 
+/// Compatibility function, used by Zig's implementation of libc in `lib/c.zig`.
+/// Assumes that `a` and `b` are multi-item pointers with NUL-termination.
```
Comment (ifreund) [inline lib/std/c.zig:43]: I think this line is unnecessary, it just restates information from the type signature in English.

File: lib/std/c.zig (hunk 1 above)
Comment (BratishkaErik) [reply lib/std/c.zig]: done

File: lib/std/c.zig (hunk 2 above)
Comment (BratishkaErik) [reply lib/std/c.zig]: done, sorry for fragmented rebase, I'm a bit confused with my branches

File: lib/std/c.zig [hunk 3]
```diff
@@ -39,6 +39,32 @@ pub fn versionCheck(comptime glibc_version: std.SemanticVersion) type {
     };
 }
 
+/// Compatibility function, used by Zig's implementation of libc in `lib/c.zig`.
+/// Returns `-1` if `a` appears before `b` in lexicographical order.
+/// Returns `1` if `a` appears after `b` in lexicographical order.
+/// Returns `0` if `a` and `b` are equal.
+/// See also `std.mem.order` for idiomatic function
+/// with parameters as Zig's slices.
+pub fn strcmp(a: [*:0]const u8, b: [*:0]const u8) i8 {
```
Comment (andrewrk) [inline lib/std/c.zig:48]: This file is supposed to have bindings to libc only. The appropriate place for a strcmp implementation is in lib/c.zig which is the actual implementation of libc functions.

I suggest to take @ifreund's suggestion to make a string comparison function that is not C related, but is having to do with two null terminated strings, put it in mem.zig and call it from both here and from lib/c.zig. Equivalent to how strlen is handled via std.mem.len.

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: lib/std/mem.zig [hunk 4]
```diff
@@ -947,6 +947,41 @@ test "len" {
...
+            .gt => .gt,
+        };
+    }
+};
+
+/// Returns `LexicographicalOrder.lt` if `a` appears before `b` in lexicographical order.
+/// Returns `LexicographicalOrder.gt` if `a` appears after `b` in lexicographical order.
+/// Returns `LexicographicalOrder.eq` if `a` and `b` are equal.
+/// See also `std.mem.order`.
+pub fn stringCompareZ(a: [*:0]const u8, b: [*:0]const u8) LexicographicalOrder {
```
Comment (andrewrk) [inline lib/std/mem.zig:968]: Please avoid the word "string" in mem functions. I don't want anyone to get the impression that this comparison is unicode-aware.

Inside `std.mem`, it is already implied that you are talking about consecutive memory regions. No need for the word "string" or any similar word.

Also, please use `std.math.Order`. The only reason for `LexicographicalOrder` to exist is the C ABI bleeding into the zig std lib, which I don't want to happen.

File: lib/std/mem.zig [hunk 5]
```diff
@@ -947,6 +947,41 @@ test "len" {
...
+    @setRuntimeSafety(false);
```
Comment (andrewrk) [inline lib/std/mem.zig:969]: why?

File: lib/std/mem.zig [hunk 6]
```diff
@@ -947,6 +947,41 @@ test "len" {
...
+    var index: usize = 0;
+    while (a[index] == b[index] and a[index] != 0) : (index += 1) {}
+    const result: i2 = @as(i2, @boolToInt(a[index] > b[index])) - @as(i2, @boolToInt(a[index] < b[index]));
+    return @intToEnum(LexicographicalOrder, result);
```
Comment (andrewrk) [inline lib/std/mem.zig:973]: Please write this without using `@intToEnum` unless you can demonstrate that it affects performance.

**Review by andrewrk — CHANGES_REQUESTED** [review]

File: lib/std/mem.zig (hunk 4 above)
Comment (BratishkaErik) [reply lib/std/mem.zig:968]: Do you mean something like `std.mem.orderZ`?

File: lib/std/mem.zig (hunk 6 above)
Comment (BratishkaErik) [reply lib/std/mem.zig:973]: > Please write this without using `@intToEnum` unless you can demonstrate that it affects performance.

Well, I compared this version with `normal` version a.k.a `return math.Order` and on my PC this version was consistently slightly faster after 1_000_000 iterations (I thought due to extra branches and stuff), but right now I'm observing that normal version is much faster on my notebook (???). Looks like it's not worth to oversmart compiler and leave it with its smart optimizations. My big thanks for the reviews!

**Review by andrewrk — APPROVED** [review]
Good work, thank you for taking the review comments in stride!

**Comment by BratishkaErik** [comment]
Thank you for reviews, they definitely improved quality a lot...

**Pull request merged**